Upgrade Docker and switch to the aufs storage driver

rtyler opened the pull request:
See [INFRA-935](https://issues.jenkins-ci.org/browse/INFRA-935) for more details.

This pull request also upgrades our garethr/docker module which we have been running a forked version of, and no longer need to do so.

Commit:
Default to using the aufs storage driver, which is not garbage

Unlike devicemapper, which is total garbage. Seeing failures in newer ci.j.io
and trusted.ci.j.io nodes where `docker push` commands are failing

Commit:
Install the linux-image-extra package to make sure we have the aufs module

Facter to the rescue! Regardless of which specific kernel version we have, at
least on Ubuntu 14.04 we want to ensure we have the appropriate package
installed

Commit:
Upgrade the garethr/docker module to what is "upstream" and set engine version to 1.12

Tested this manually with Vagrant/serverspec by provisioning a machine with the
current production branch and then re-running Puppet with this change

The only potential issue seen was that switching from devicemapper to aufs
requires the daemon to be restarted before any pulls can happen. This might make
sense to do as a manual step as needed.

batmat commented on dist/profile/manifests/docker.pp:
```diff
@@ -2,10 +2,11 @@
   class { '::docker':
-    version       => '1.9.1',
     # Disabling the management of the kernel, since we have to pre-install
     # kernel modules on Ubuntu 12.04 LTS and restart the host machine anyways
```
So from what you say on IRC, that comment above isn't actual anymore right?

rtyler replied:
I'll replace 12.04 with 14.04, because it actually still applies :P

batmat reviewed (approved):
\o/ :+1:  FWIU

Commit:
Correct the comment in profile::docker

The aufs module is still not in the default images, so we must manage
linux-image-extra separately
